src/Mono.Android/Android.Runtime/JNIEnv.cs

```diff
@@ -162,33 +164,51 @@ static unsafe void RegisterJniNatives (IntPtr typeName_ptr, int typeName_len, In
 			if (methods.Length == 0)
 				return;
 
-			JniNativeMethod[] natives = new JniNativeMethod [methods.Length-1];
-			for (int i = 0; i < methods.Length; ++i) {
-				string method = methods [i];
-				if (string.IsNullOrEmpty (method))
-					continue;
-				string[] toks = methods [i].Split (new[]{':'}, 4);
-				Delegate callback;
-				if (toks [2] == "__export__") {
-					var mname = toks [0].Substring (2);
-					var minfo = type.GetMethods (BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static).Where (m => m.Name == mname && JavaNativeTypeManager.GetJniSignature (m) == toks [1]).FirstOrDefault ();
-					if (minfo == null)
-						throw new InvalidOperationException (String.Format ("Specified managed method '{0}' was not found. Signature: {1}", mname, toks [1]));
-					callback = CreateDynamicCallback (minfo);
-				} else {
-					GetCallbackHandler connector = (GetCallbackHandler) Delegate.CreateDelegate (typeof (GetCallbackHandler), 
-						toks.Length == 4 ? Type.GetType (toks [3], true) : type, toks [2]);
-					callback = connector ();
+			bool registered = false;
+#if JAVA_INTEROP
+			if (androidRuntime != null) {
+				JniType jniType = null;
+				JniType.GetCachedJniType (ref jniType, Java.Interop.TypeManager.GetClassName (jniClass));
+
+				if (jniType != null) {
+					try {
+						androidRuntime.TypeManager.RegisterNativeMembers (jniType, type, null);
+						registered = true;
+					} catch (NotSupportedException e) {
+						// ignore, lets fallback bellow
+					}
 				}
-				natives [i] = new JniNativeMethod (toks [0], toks [1], callback);
 			}
+#endif
+			if (!registered) {
+				JniNativeMethod[] natives = new JniNativeMethod [methods.Length-1];
```

Comment:
...as such, I'd prefer that *this* block of code be *moved* into `AndroidTypeManager`, in `AndroidRuntime.cs`:

```csharp
partial class AndroidTypeManager : JniRuntime.JniTypeManager {
    public override void RegisterNativeMembers (JniType nativeClass, Type type, string methods)
    {
        try {
            base.RegisterNativeMembers (nativeClass, type, methods);
            return;
        } catch (NotSupportedException) {
            // ignore; legacy fallback is below
        }
        string[] members = methods.Split ('\n');
        JniNativeMethod[] natives = new JniNativeMethod [members.Length-1];
        // ...
    }
}
```

Meanwhile, the callsite within `JNIEnv.RegisterJniNatives()` would be:

```csharp
string methods = methods_ptr == IntPtr.Zero ? null : new string ((char*) methods_ptr, 0, methods_len);
androidRuntime.TypeManager.RegisterNativeMembers (jniType, type, methods)
```

Doing this "unifies" JNI registration logic to be in one location -- `AndroidTypeManager` -- instead of in two places (`AndroidTypeManager` and `JNIEnv`).